File: atcoder/arc032/b.py
union_find_parent = {}

def find(x):
    if union_find_parent[x] != x:
        union_find_parent[x] = find(union_find_parent[x])
    return union_find_parent[x]

def union(x, y):
    union_find_parent[find(x)] = find(y)


(N, M) = [int(x) for x in input().split()]
for i in range(N):
    union_find_parent[i+1] = i+1

for _ in range(M):
    (a, b) = [int(x) for x in input().split()]
    union(a, b)

cnt = 0
for k, v in union_find_parent.items():
    if k == v:
        cnt += 1
print(cnt-1)

File: atcoder/arc032/b.js
function UnionFind(size){
  this._values = new Array(size);
  this._values.fill(-1);
}

UnionFind.prototype = {
  find: function(x){
    while(this._values[x] >= 0){
      x = this._values[x];
    }
    return x;
  },
  union: function(x,y){
    px = this.find(x);
    py = this.find(y);
    if (px !== py){
      if(this._values[px] > this._values[py]){
        this._values[px] += this._values[py];
        this._values[py] = px;
      }else{
        this._values[py] += this._values[px];
        this._values[px] = py;
      }
    return this._values[px];
    }
  },
  count: function(){
    var group = 0;
    for(var i = 0; i<this._values.length; ++i){
      if(this._values[i] < 0){
        group += 1;
      }
    }
    return group;
  }
};


function main(input){
  var rows = input.split("\n");
  var n,m;
  n = parseInt(rows[0].split(" ")[0],10);
  m = parseInt(rows[0].split(" ")[1],10);
  var uf = new UnionFind(n);
  for(var i=0; i<m; i++){
    var a,b;
    a = parseInt(rows[i+1].split(" ")[0], 10) - 1;
    b = parseInt(rows[i+1].split(" ")[1], 10) - 1;
    uf.union(a,b);
  }
  console.log(uf.count()-1);
}

main(require("fs").readFileSync("/dev/stdin", "utf8"));
